[WIP NOT for 0.13.0] Quick guide for Codewind on Minikube

j-c-berger opened the pull request:
Signed-off-by: Jacob Berger <jacob.berger@ibm.com>

Sample for quick-guide structure. 

For issue #2407: https://github.com/eclipse/codewind/issues/2407

j-c-berger commented:
@johnmcollier, please see this PR for quick guide on minikube.

johnmcollier commented on docs/_documentations/codewind-eclipse-quick-guide.md:
```diff
@@ -0,0 +1,170 @@
+---
+layout: docs
```
This doc seems to be for local Codewind, do we need it for the Minikube guide?

j-c-berger replied:
Good point. Since quick guides are new, they don't have their own layout yet like `docs` and `guides` do. So, let me ask the larger ID team. I believe we want it laid out more like a doc, since it's quicker and shorter than a guide, but I'll confirm!

johnmcollier commented on docs/_documentations/codewind-minikube-guickguide.md:
```diff
@@ -0,0 +1,110 @@
+    * **Password:** In step 3, developer password given to cwctl in the value of `kdevpass`.
+5. To save and create the connection to the Codewind instance on Minikube, click `Save`. 
+
+## Developing a microservice
```
We can re-use the developing a microservice section from the local guide here

j-c-berger replied:
Are you talking about the [Getting started with Codewind in VS Code guide](https://kabanero.io/guide/codewind-getting-started-vscode/) and [Getting started with Codewind in Eclipse guide](https://kabanero.io/guide/codewind-getting-started-eclipse/)? 

The instructions for developing a microserice differ between VS Code and Eclipse, so they won't be the same, pending on the IDE users work in. Maybe we could link to them, that would defeat the purpose of a quick, self-contained guide though. 

In my new commit, I've added the instructions for developing a microservice within VS Code. Maybe we can tailor it more to Minikube? Right now, it's solely developing a microservice, using Codewind on VS Code.

johnmcollier commented on docs/_documentations/codewind-vscode-quick-guide.md:
```diff
@@ -0,0 +1,166 @@
+order: 1
+---
+
+# Getting Started with Codewind in VS Code
```
Same as for the eclipse guide, do we need this for the minikube guide since it's local only?

j-c-berger commented:
@johnmcollier, I've made a new commit, see ```added microservice instructions``` addressing your comments. Please see my responses to your comments, since some I'm not yet sure of.

johnmcollier commented on docs/_documentations/codewind-minikube-guickguide.md:
```diff
@@ -0,0 +1,181 @@
+1. Under the Explorer pane, select `Codewind`.
+2. Expand `Codewind` by clicking the drop-down arrow.
+3. Hover over the `Projects` entry underneath Codewind in the Explorer pane, and press the `+` icon to create a project.
+    * **Note:** Make sure Docker is running. Otherwise, you get an error.
```
We can remove this section, as Docker isn't applicable here

johnmcollier commented on docs/_documentations/codewind-minikube-guickguide.md:
```diff
@@ -0,0 +1,181 @@
+
+You now can work with your calculator.
+
+1. Use the port number you saw when you first opened the application.
```
Change this to something like "Use the URL you saw when you first opened the application"

johnmcollier commented on docs/_documentations/codewind-minikube-guickguide.md:
```diff
@@ -0,0 +1,181 @@
+
+1. Use the port number you saw when you first opened the application.
+2. Make sure to remove the `< >` symbol in the URL.
+3. `http://127.0.0.1:<port>/starter/calculator/aboutme`
```
3. `<url>/starter/calculator/aboutme`

johnmcollier commented on docs/_documentations/codewind-minikube-guickguide.md:
```diff
@@ -0,0 +1,181 @@
+
+You could also try a few of the sample calculator functions:
+
+* `http://127.0.0.1:<port>/starter/calculator/{op}/{a}/{b}`, where you can input one of the available operations `(+, _, *)`, and an integer a, and an integer b.
```
Same URL changes here as mentioned above

j-c-berger commented:
@johnmcollier, thanks. I'll get a new commit up shortly. Thanks for testing it as well.

jgwest commented on docs/_documentations/codewind-minikube-guickguide.md:
```diff
@@ -0,0 +1,110 @@
+    * Follow the install instructions for [Eclipse](https://www.eclipse.org/codewind/mdt-eclipse-getting-started.html) or [VS Code](https://www.eclipse.org/codewind/mdt-vsc-getting-started.html). 
+2. Locate the `cwctl` command line interface (CLI). 
+    * On Mac and Linux, the cwctl is `~/.codewind/<version>/cwctl`. 
+    * On Windows, the cwctl is located: ??? 
```
👍

j-c-berger commented:
@johnmcollier, thanks for the talk today. Per our conversation, new commit, `revised microservice instructions` is up. Thanks!

j-c-berger commented:
@johnmcollier new commit is up. Please review the guide. 

@naveenkaratekid, please do a test of the guide to ensure accuracy of steps and content. 

Thank you both and don't hesitate to let me know what I must do!

naveenkaratekid commented on docs/_documentations/codewind-minikube-guickguide.md:
````diff
@@ -0,0 +1,198 @@
+        --ingress <ingress-ip>.nip.io
+        ```
+4. To install cwctl on Windows, run: `cwctl.exe --insecure install remote --namespace codewind --kadminuser admin --kadminpass admin --krealm codewind --kclient codewind --kdevuser developer --kdevpass developer --ingress <ingress-ip>.nip.io`     
+    * **Note:** The `ingress-ip` value is the value you found in Step 4 of `Preparing Minikube`.
````
When I ran step 3 for installing remote in cwctl, it seems to be taking forever.

I am seeing this as the output every time:
```WARN[1650] Watch for Pod codewind-keycloak-ka5ft7nq ended. Timeout or connection error. 
INFO[1650] Waiting for pod: codewind-keycloak-ka5ft7nq  
INFO[1650] codewind-keycloak-ka5ft7nq, phase: Pending   
INFO[1650] codewind-keycloak-ka5ft7nq, phase: Pending ContainersNotReady  ```

I have tried to delete and restart minikube, and run cwctl stop-all to see if that helps, but it didn't.

Even though this is a quick guide, I don't seem to see any troubleshooting tips for this.

johnmcollier replied:
@naveenkaratekid Can you try following the instructions for the Codewind operator and seeing if they work for you? Wondering if the quicklab should use the operator isntead

j-c-berger replied:
@johnmcollier what instructions for the Codewind operator are you referring to?

j-c-berger replied:
Do you mean the instructions laid out here in [eclipse/codewind-operator](https://github.com/eclipse/codewind-operator)?

naveenkaratekid replied:
@johnmcollier I went through the steps in the codewind operator, and when I ran the install script with the ingress IP as the argument, this is what I got:

```
Naveens-MacBook-Pro:deploy naveenk@ibm.com$ ./install.sh operator -i 192.168.64.12.nip.io
############################
Codewind Operator install.sh
############################

########################
   Codewind Operator
########################

Ingress Domain: 192.168.64.12.nip.io
Target Openshift 311: false
Creating Codewind namespace:
Error from server (AlreadyExists): namespaces "codewind" already exists
Deploying Operator Service Account:
Error from server (AlreadyExists): error when creating "service_account.yaml": serviceaccounts "codewind-operator" already exists
Deploying Operator RBAC Roles:
Error from server (AlreadyExists): error when creating "role.yaml": roles.rbac.authorization.k8s.io "codewind-operator" already exists
Deploying Operator RBAC Role Bindings:
Error from server (AlreadyExists): error when creating "role_binding.yaml": rolebindings.rbac.authorization.k8s.io "codewind-operator" already exists
Deploying Operator Cluster Roles:
Error from server (AlreadyExists): error when creating "cluster_roles.yaml": clusterroles.rbac.authorization.k8s.io "codewind-operator-cluster" already exists
Deploying Codewind Cluster Role Bindings:
Error from server (AlreadyExists): error when creating "cluster_role_binding.yaml": clusterrolebindings.rbac.authorization.k8s.io "codewind-operator-clb" already exists

Installing Custom Resource Definitions (CRD):
Error from server (AlreadyExists): error when creating "codewind.eclipse.org_keycloaks_crd.yaml": customresourcedefinitions.apiextensions.k8s.io "keycloaks.codewind.eclipse.org" already exists
Error from server (AlreadyExists): error when creating "codewind.eclipse.org_codewinds_crd.yaml": customresourcedefinitions.apiextensions.k8s.io "codewinds.codewind.eclipse.org" already exists
Creating Codewind configmap:
Error from server (AlreadyExists): error when creating "custom-codewind-configmap.yaml": configmaps "codewind-operator" already exists
Deploying Codewind operator:
Error from server (AlreadyExists): error when creating "operator.yaml": deployments.apps "codewind-operator" already exists
Requesting a new Keycloak service
Error from server (AlreadyExists): error when creating "codewind.eclipse.org_v1alpha1_keycloak_cr.yaml": keycloaks.codewind.eclipse.org "devex001" already exists
Reading Keycloak deployments
NAME       NAMESPACE   AGE   ACCESS
devex001   codewind    17s   
Waiting for keycloak (may take a few minutes Pending->ContainerCreating->Running)
keycloak 
```

micgibso replied:
Moving this guide out to the next release.

johnmcollier replied:
@naveenkaratekid Can you try the following?
```
$ minikube stop
$ minikube delete
$ minikube start --memory=4gb
```

Then reproduce the operator install steps

j-c-berger replied:
@johnmcollier, I heard from Naveen that these instructions don't work either. Any other thoughts on what we can do to get Codewind installed on Minikube?

j-c-berger commented on docs/_documentations/codewind-minikube-guickguide.md:
```diff
@@ -0,0 +1,198 @@
+## Installing Codewind
+1. Choose your IDE and install the Codewind extension. 
+    * Follow the install instructions for [Eclipse](https://www.eclipse.org/codewind/mdt-eclipse-getting-started.html) or [VS Code](https://www.eclipse.org/codewind/mdt-vsc-getting-started.html). 
+2. Locate the `cwctl` command line interface (CLI). 
```
let's use the Codewind operator instructions instead of `cwctl` instructions. Let's simplify the Codewind operator steps include them here in lieu of `cwctl`. 

If we want to link back to the docs, we can link users here: https://www.eclipse.org/codewind/remote-deploying-codewind.html

micgibso commented:
In discussion with Toby, this is not a supported environment so closing this issue.
